pkg/pool-weighted/contracts/smart/ManagedPool.sol

```diff
@@ -349,6 +352,100 @@ contract ManagedPool is BaseWeightedPool, ReentrancyGuard {
+     * - Deregister the token in the Vault (since it now has zero balance)
+     * - Reduce the total weight by the weight of the token removed, and update the number of tokens and `_tokenState`
+     *   (all other weights then scale accordingly)
+     */
+    function removeToken(uint256 tokenIndex, address recipient) external authenticate whenNotPaused returns (uint256) {
+        return _removeToken(tokenIndex, recipient);
+    }
+
+    function _removeToken(uint256 tokenIndex, address recipient) internal returns (uint256) {
+        // Calculate the BPT value of the full token balance (other values returned to avoid duplication)
+        //(uint256 bptAmountIn, uint256 minAmountOut, uint256 normalizedWeight) = getTotalBptForToken(tokenIndex);
+        // Exit the pool, returning the full balance of the token to the recipient
+        (IERC20[] memory tokens, uint256[] memory rawBalances, ) = getVault().getPoolTokens(getPoolId());
+        _require(tokenIndex < tokens.length, Errors.OUT_OF_BOUNDS);
+        _require(tokens.length > 2, Errors.MIN_TOKENS);
+        // Verify there is no ongoing weight change
+        _require(
+            0 == _calculateWeightChangeProgress() || FixedPoint.ONE == _calculateWeightChangeProgress(),
+            Errors.REMOVE_TOKEN_DURING_WEIGHT_CHANGE
+        );
+
+        IERC20 removedToken = tokens[tokenIndex];
+        uint256 weightOfRemovedToken = _exitWithEntireBalance(
+            tokens,
+            _getNormalizedWeights(),
+            tokenIndex,
+            _scalingFactor(removedToken),
+            rawBalances[tokenIndex],
+            recipient
+        );
+
+        // Deregister the token in the Vault
+        IERC20[] memory tokensToRemove = new IERC20[](1);
+        tokensToRemove[0] = removedToken;
+        getVault().deregisterTokens(getPoolId(), tokensToRemove);
+
+        emit TokenRemoved(removedToken);
+
+        // Clean up data structures and update the token count
+        delete _tokenState[removedToken];
+        _setMiscData(_getMiscData().insertUint7(tokens.length - 1, _TOTAL_TOKENS_OFFSET));
+
+        // Decrease the total weight by the weight of the token being removed
+        _totalWeight -= weightOfRemovedToken;
+
+        // The bpt amount corresponding to the token is simply its proportional share of the totalSupply
+        return weightOfRemovedToken.mulDown(totalSupply());
+    }
+
+    // Separated to avoid stack too deep
+    function _exitWithEntireBalance(
+        IERC20[] memory tokens,
+        uint256[] memory normalizedWeights,
+        uint256 tokenIndex,
+        uint256 scalingFactor,
+        uint256 unscaledTokenAmountOut,
+        address recipient
+    ) private returns (uint256) {
+        uint256[] memory minAmountsOut = new uint256[](tokens.length);
+        // In the limit, compensate for BasePool downscaling of amountOut
+        minAmountsOut[tokenIndex] = _downscaleUp(unscaledTokenAmountOut, scalingFactor);
```

Comment:
Limits are checked in the Vault, where there is no scaling. So we shouldn't have to change anything here, and we specially shouldn't downscale an unscaled amount (we only downscale upscaled amounts).